keps/sig-network/2079-network-policy-port-range/README.md

```diff
@@ -0,0 +1,306 @@
+proposes to add a new field that allows a declaration of a port range, 
+simplifying the creation of rules with multiple ports.
+
+## Motivation
+
+NetworkPolicy object is a complex object, that allows a developer to specify 
+what's the traffic behavior expected of the application and allow/deny 
+undesired traffic.
+
+There are a number of user issues like [kubernetes #67526](https://github.com/kubernetes/kubernetes/issues/67526) 
+and [kubernetes #93111](https://github.com/kubernetes/kubernetes/issues/93111) 
+where users expose the need to create a policy that allow a range of ports but some 
+specific port, or also cases that a user wants to create a policy that allows 
+the egress to other cluster to the NodePort range (eg 32000-32768) and in this case, 
+the rule should be created specifying each port separately, as:
+
+```
+spec:
+  egress:
+  - ports:
+    - protocol: TCP
+      port: 32000
+    - protocol: TCP
+      port: 32001
+    - protocol: TCP
+      port: 32002
+    - protocol: TCP
+      port: 32003
+[...]
+    - protocol: TCP
+      port: 32768
+```
+
+So for the user:
+* To allow a range of ports, each of them must be declared as an item from 
+``ports`` array
+* To make an exception needs a declaration of all ports but the exception
+
+Adding a new ``PortRange`` field inside the ``ports`` will allow a simpler 
+creation of NetworkPolicy to the user.
+
+### Goals
+
+Add Range field to Ports in NetworkPolicy
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+In NetworkPolicy specification, inside ``NetworkPolicyPort`` object struct 
+specify a new ``Range`` field composed of the minimum and maximum ports 
+inside the range.
+
+If both ``Port`` and ``Range`` are specified, they are cumulative and no further
+validation might occur. It's up to the CNI to summarize both of the fields in a 
+single rule.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I have an application that communicates with NodePorts of a different cluster 
+and I want to allow the egress of the traffic only the NodePort range 
+(eg. 30000-32767) as I don't know which port is going to be allocated on the 
+other side, but don't want to create a rule for each of them.
+
+### Notes/Constraints/Caveats
+
+*  The technology used by the CNI provider might not support port range in a 
+trivial way as described in [#drawbacks]
+
+### Risks and Mitigations
+
+CNIs will need to support the new field in their controllers. For this case 
+we'll try to make broader communication with the main CNIs so they can be aware
+of the new field.
+
+## Design Details
+
+API changes to NetworkPolicy:
+* Add a new struct called ``NetworkPolicyPortRange`` as the following:
+```
+type NetworkPolicyPortRange struct {
+    From         uint16
+    To           uint16
+    Exceptions:  []uint16
+}
+```
+* Add a new field ``spec.ingress|egress.ports.range`` that points to the
+new struct:
+```
+// NetworkPolicyPort describes a port to allow traffic on
+type NetworkPolicyPort struct {
```

Comment:
Can we add the whole struct here instead of just the range field?

Reply:
Added. BTW the documentation says:

```
// The port on the given protocol. This can either be a numerical or named port on
	// a pod. If this field is not provided, this matches all port names and numbers.
	// +optional
	Port *intstr.IntOrString
```

I'm worried here about if we provide ``Range`` but no ``Port`` it will still match all Ports, so I've changed the struct doc to:

```
// The port on the given protocol. This can either be a numerical or named 
  // port on a pod. If this field is not provided but a Range is 
  // provided, this field is ignored. Otherwise this matches all port names and
  // numbers.
```

This validation will probably need to be changed in the CNI? Or I need to check if this is something from the API

Reply:
How about:

```
  // The port on the given protocol. This can either be a numerical or named 
  // port on a pod. If this field is not provided, this matches all port names and
  // numbers, whether an endPort is defined or not.
```

Or were you actually proposing something different?

Reply:
IMO it's good. We're exactly proposing in validations (I think we are!) that endPort can only exist if Port is defined.

Also I think this comment is old, when we were still considering using PortRange (and not EndPort), the new comment was:

```
// The port on the given protocol. This can either be a numerical or named port on
	// a pod. If this field is not provided, this matches all port names and numbers.
```

But I've changed as you've proposed, it's clearer